### PR TITLE
stm32f2: Add DMA driver

### PR DESCRIPTION
**Pull request by AurelienGONCE**
DMA peripheral has been created to give only one point of DMA configuration and transfer. 

Before, the DMA was directly configured in UART. The need of the DMA to be used with SPI induces to separate in a different peripheral.

At this time, the DMA is used in UART and in SPI (only spi_transfer_bytes). The DMA interruptions are managed directly by the DMA peripheral. 

2 mutual exclusions (mutex) are used in DMA:
- the first one in configuration to avoid data corruption during the transfer,
- the last one to wait the end of DMA transmission.

Tests are done with UART and SPI operating on the same stream and all communications succeeded but without compiler optimization. During DMA transfer, the hand is given to other thread.

All others stm32 cpu are a little bit impacted. The defines used to call common spi functions was in common_cpu.h and have been moved to each corresponding files of each stm32 cpu. In stm32f2, the spi_transfer_bytes function has been implemented to use DMA peripheral and the common function is not used anymore.

How to use DMA:
- DMA initialization during the initialization of peripheral (UART, SPI ...),
- Acquire configuration mutex of the corresponding stream DMA,
- Configure DMA,
- Enable the DMA to run,
- Wait until the end of transmission in acquiring the transmission mutex,
  (When the transmission is done, the corresponding interrupt is called. The transmission mutex is released and the hand is given back to the thread)
- Disable the DMA (make it before "Relase the configuration mutex", even it could result to make a partial transmission),
- Release the configration mutex to give the resource available for another peripheral.

### CODE REVIEW

**Comment by toonst** [comment]
This is based on https://github.com/RIOT-OS/RIOT/pull/5397

**Comment by AurelienGONCE** [comment]
sorry but what I have to do with this?
________________________________
De : Toon Stegen [notifications@github.com]
Envoyé : mercredi 29 juin 2016 14:06
À : RIOT-OS/RIOT
Cc : Gonce Aurélien; Author
Objet : Re: [RIOT-OS/RIOT] Pr/dma (#5584)


This is based on #5397<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_RIOT-2DOS_RIOT_pull_5397&d=CwMCaQ&c=cxWN2QSDopt5SklNfbjIjg&r=LOLWya72aM4nx80lbDE8RHEyL7wTYqOzvPbY9SId7ck&m=3Qjp-JVB7A0nPh8mvqNjWbcPAb39GEku_ZqFZeIZ7Ts&s=e0hclEqhh9-F4KMYCdRnUO8IUgrvkfyBHWrdUYLtHSM&e=>

**Comment by AurelienGONCE** [comment]
We found an issue when using compiler optimization. 

We have investigated and found that in "dma_stream_config" function, the parameter "uint32_t periph_addr_reg" is optimized by the compiler. This parameter is used to give the peripheral data register and transmit data through the peripheral. The address is not transmitted correctly (0 is passed on) and It results a failed in DMA transmission.

A volatile data (data register of UART, SPI ...) should be transmitted through this parameter and give it to an other volatile data (DMA stream x peripheral address register) in "dma_stream_config" function. 

At this time, the only solution we found, it is to create a volatile local data to transmit the data register: `volatile uint32_t periph_addr = periph_addr_reg;`

We are not so happy with this solution, so if you have any other idea, we are listening.

**Comment by gebart** [comment]
Can this DMA driver be used by other stm32 CPUs? Is it possible to move to the stm32_common part of the CPU tree?

**Comment by AurelienGONCE** [comment]
As per what I found on Internet, it seems that there are some differences in DMA for each CPU. Several application note can be found regrouping F2, F4 and F7 together, F1 and L1 together and one for only F0. Possibly the only difference between the cpu families is the number of channels/streams, so only some specific code for interrupt and register management would be needed.

At this time, we are only working on STM32F2 and we don't have resources to make the code for each stm32 CPUs and to test it.

This PR could be the starting point for something more generic for who wants and can to do it.

**Comment by vincent-d** [comment]
Rebased on latest #5397

**Comment by pwillem** [comment]
This pr now depends on #5842

**Comment by vincent-d** [comment]
dma.h header is not generic, I moved it to the cpu folder.

Hopefully, this PR is ready for review :)

**Comment by DipSwitch** [comment]
> We found an issue when using compiler optimization.
> 
> We have investigated and found that in "dma_stream_config" function, the parameter "uint32_t periph_addr_reg" is optimized by the compiler. This parameter is used to give the peripheral data register and transmit data through the peripheral. The address is not transmitted correctly (0 is passed on) and It results a failed in DMA transmission.
> 
> A volatile data (data register of UART, SPI ...) should be transmitted through this parameter and give it to an other volatile data (DMA stream x peripheral address register) in "dma_stream_config" function.
> 
> At this time, the only solution we found, it is to create a volatile local data to transmit the data register: volatile uint32_t periph_addr = periph_addr_reg;
> 
> We are not so happy with this solution, so if you have any other idea, we are listening.

Have you tried to pass it as a pointer? So instead of `void dma_stream_config(dma_t stream_dev, uint32_t periph_addr_reg, uint32_t dma_config, char* data, uint16_t length)` use `void dma_stream_config(dma_t stream_dev, void* periph_addr, uint32_t dma_config, char* data, uint16_t length)` maybe this could make the compiler not make the optimization's it would make for `uint32_t`'s

File: cpu/stm32f2/periph/spi.c [hunk 1]
```diff
@@ -397,28 +429,142 @@ void spi_poweroff(spi_t dev)
...
+int spi_transfer_bytes(spi_t dev, char *out, char *in, unsigned int length)
 {
+    SPI_TypeDef *spi_dev;
+    uint32_t tx_stream;
+    uint32_t rx_stream;
+    uint32_t tx_channel;
+    uint32_t rx_channel;
+    uint32_t tx_conf;
+    uint32_t rx_conf;
+    uint32_t periph_addr_reg;
```
Comment (DipSwitch) [inline cpu/stm32f2/periph/spi.c]: Try to make it a `void*`

File: cpu/stm32f2/periph/spi.c [hunk 2]
```diff
@@ -397,28 +429,142 @@ void spi_poweroff(spi_t dev)
...
+            rx_stream = SPI_2_DMA_RX_STREAM;
+            tx_channel = SPI_2_DMA_TX_CHANNEL;
+            rx_channel = SPI_2_DMA_RX_CHANNEL;
+            break;
+#endif  /* SPI_2_EN */
+        default:
+            return -1;
+    }
+
+    periph_addr_reg = (uint32_t)&(spi_dev->DR);
```
Comment (DipSwitch) [inline cpu/stm32f2/periph/spi.c]: Don't use a cast.

File: cpu/stm32f2/periph/spi.c [hunk 3]
```diff
@@ -397,28 +429,142 @@ void spi_poweroff(spi_t dev)
...
+        tx_conf = ((tx_channel << 25) | DMA_SxCR_PL_0 | DMA_SxCR_DIR_0 | DMA_SxCR_MINC | DMA_SxCR_TCIE);
+    }
+
+    /* acquire dma spi rx and tx streams */
+    dma_conf_acquire(tx_stream);
+    dma_conf_acquire(rx_stream);
+
+    /* streams configuration */
+    dma_stream_config(tx_stream, periph_addr_reg, tx_conf, out, (uint16_t)length);
+    dma_stream_config(rx_stream, periph_addr_reg, rx_conf, in, (uint16_t)length);
```
Comment (DipSwitch) [inline cpu/stm32f2/periph/spi.c]: Try and use modified function prototype. I think GCC's `LTO` is screwing this up :)

File: cpu/stm32f2/periph/uart.c [hunk 4]
```diff
@@ -133,65 +124,102 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
+        uint32_t tx_conf = ((uart_config[uart].dma_chan << 25) |
+                            DMA_SxCR_PL_0 |
+                            DMA_SxCR_MINC |
+                            DMA_SxCR_DIR_0 |
+                            DMA_SxCR_TCIE);
+
+        /* acquire dma uart stream */
+        dma_conf_acquire(uart_config[uart].dma_stream);
+        /* UART stream configuration */
+        dma_stream_config(uart_config[uart].dma_stream, (uint32_t)&(dev->DR),
```
Comment (DipSwitch) [inline cpu/stm32f2/periph/uart.c]: Same goes here off course :)

File: cpu/stm32f2/periph/spi.c (hunk 3 above)
Comment (DipSwitch) [inline cpu/stm32f2/periph/spi.c]: Oh, although LTO is not enabled per default yet I see, still worth a try however =)

**Comment by PeterKietzmann** [comment]
I'm sorry to say but the stm32 disco board make use of one common SPI driver now.

**Comment by vincent-d** [comment]
This PR is outdated. We worked on DMA with old drivers. SPI and UART are now common.

I'll close that one.

**Pull request closed without merging**